python/mxnet/gluon/block.py

```diff
@@ -1094,8 +1094,57 @@ def _deferred_infer_shape(self, *args):
             raise ValueError(error_msg)
 
     def _call_cached_op(self, *args):
-        if self._cached_op is None:
+        if not self._cached_op_args or not self._cached_graph:
             self._build_cache(*args)
+
+        if self._first_forward:
+            # partition static shape ops if the graph contains any dynamic shape op
+            self._first_forward = False
+            data, out = self._cached_graph
+            out, is_dynamic = out._optimize_for_dynamic_shape_op(is_np_array(), self._flags)
+            self._cached_graph = data, out
```

Comment:
We cant override the `self._cached_graph` with the dynamic shape partitioned graph since users can call `optimize_for` at any time after the model is hybridized. We dont want to have users to try and optimize the "dynamic shape graph partitioned" version. 

Are you planning on pushing this down into the creation of the C++ CachedOp? or some other approach to resolve this issue?

Reply:
Right, we forgot to remove this line after the dynamic shape partitioning. Will address it.

Reply:
We are still evaluating moving this logic to C++ CachedOp vs just reusing optimize_for here. In the latter case too, not updating self._cached_graph after dynamic shape partitioning should work.

Reply:
As discussed offline, we are going forward with `optimize_for` and not moving the logic to inside C++ CachedOp.